Registrar wait up to 60 seconds for SPIRE Server to start

faisal-memon opened the pull request:
Signed-off-by: Faisal Memon <fymemon@yahoo.com>

<!--
1. If this is your first PR, please read our contributor guidelines
https://github.com/spiffe/spiffe/blob/master/CONTRIBUTING.md
https://github.com/spiffe/spire/blob/main/CONTRIBUTING.md

2. Please remember to include a DCO on every commit (`git commit -s`)
https://github.com/apps/dco
-->

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [ ] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**
k8s-workload-registrar

**Description of change**
Has the k8s-workload-registrar wait up to 60 seconds for the spire server to be up before restarting. This is to support spire-server with large configs that takes a while to start up.

**Which issue this PR fixes**
fixes #2872

azdagron commented on support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:
```diff
@@ -91,7 +92,13 @@ func (e *SVID) MintSVID(ctx context.Context, key crypto.Signer) (err error) {
+		Factor:   2.0,
+		Jitter:   0.1,
+	}
+	err = retry.OnError(backoff, e.mintSVIDRetry, func() (err error) {
```
How long was the backoff for previous to this change? One minute feels like a long time to detect a misconfiguration...

faisal-memon replied:
about 1.5 seconds. On slack someone mentioned needing up to a minute because they had a log of OPA rules to process, so thats where the 1 minute number comes from. It is a bit long, i agree.

azdagron commented:
For posterity, we discussed this PR during a contributor sync up and the outcome was that 1 minute was too long. The consensus was to wait five seconds and that logging should be added so that it is clear attempts are being made.

faisal-memon commented:
> For posterity, we discussed this PR during a contributor sync up and the outcome was that 1 minute was too long. The consensus was to wait five seconds and that logging should be added so that it is clear attempts are being made.

Updated with a 5 second timeout and logs. I also added a hidden configurable to increase the timeout to cover the use cases where it takes spire server a long time to start.

azdagron commented:
Thanks for updating, @faisal-memon!

>  I also added a hidden configurable

It has been the experience on this project, that hidden (i.e. undocumented) configurables always become documented over time. In my opinion, I think we should either hard code the number of retries or we document the configurable. Curious what the other @spiffe/spire-maintainers think.

In terms of the configurable, # of attempts is hard for somebody to guess unless they understand how the backoff works. I think it would be more ergonomic to configure the max duration to wait and configure the backoff accordingly.

faisal-memon commented:
> Thanks for updating, @faisal-memon!
> 
> > I also added a hidden configurable
> 
> It has been the experience on this project, that hidden (i.e. undocumented) configurables always become documented over time. In my opinion, I think we should either hard code the number of retries or we document the configurable. Curious what the other @spiffe/spire-maintainers think.
> 
> In terms of the configurable, # of attempts is hard for somebody to guess unless they understand how the backoff works. I think it would be more ergonomic to configure the max duration to wait and configure the backoff accordingly.

makes sense, ill remove the configurable.

azdagron reviewed (approved):
\o/

edwarnicke commented:
@azdagron Any idea when this will be in a release version of Spire?

azdagron commented:
Great timing! The commit for the 1.3.0 release is likely getting picked today (we have one outstanding PR), which means 1.3.0 is expected to be released one week from today, which is April 10th.
